app/lib/service/services.dart

```diff
@@ -123,15 +125,34 @@ Future<void> withFakeServices({
     register(#appengine.context, FakeClientContext());
     registerDbService(RetryDatastoreDB(DatastoreDB(datastore!)));
     registerStorageService(storage!);
+    IOServer? frontendServer;
     if (configuration == null) {
       // start storage server
       final storageServer = FakeStorageServer(storage);
       await storageServer.start();
       registerScopeExitCallback(storageServer.close);
 
+      frontendServer = await IOServer.bind('localhost', 0);
+      final frontendServerUri =
+          Uri.parse('http://localhost:${frontendServer.server.port}');
+      registerScopeExitCallback(frontendServer.close);
+
       // update configuration
       configuration = Configuration.test(
-          storageBaseUrl: 'http://localhost:${storageServer.port}');
+        storageBaseUrl: 'http://localhost:${storageServer.port}',
+        primaryApiUri: frontendServerUri,
+        primarySiteUri: frontendServerUri,
+      );
+      serveRequests(frontendServer.server, (rq) async {
+        return await withFakeServices(
```

Comment:
Just curious, why do we need to call `withFakeServices` again?

Reply:
The HTTP server's request processing would be in a new top-level scope, we need to re-initialize everything. It is possible that we could get the same result without it, but it may need tweaking how we subscribe to the server.

Reply:
But are we creating new instances of fake services... couldn't that have negative side-effects for stateful fake services?

Reply:
The stateful parts (datastore and storage) are passed in, the rest is mostly cheap object creation that lives only for the request scope and will be discarded.

I also wish that we wouldn't need this, so I'll add a TODO and revisit it later.

Reply:
great, because I plan on adding more stateful services :see_no_evil: